lib/cast/int32.js

```diff
@@ -0,0 +1,43 @@
+'use strict';
+
+const assert = require('assert');
+const BSON = require('bson');
+
+/**
+ * Given a value, cast it to a Int32, or throw an `Error` if the value
+ * cannot be casted. `null` and `undefined` are considered valid.
+ *
+ * @param {Any} value
+ * @return {Number}
+ * @throws {Error} if `value` does not represent an integer, or is outside the bounds of an 32-bit integer.
+ * @api private
+ */
+
+module.exports = function castInt32(val) {
+  if (val == null) {
+    return val;
+  }
+  if (val === '') {
+    return null;
+  }
+
+  let coercedVal;
+  if (val instanceof BSON.Int32 || val instanceof BSON.Double) {
+    coercedVal = val.value;
+  } else if (val instanceof BSON.Long) {
+    coercedVal = val.toNumber();
+  } else {
+    coercedVal = Number(val);
+  }
```

Comment:
What is the type detection for Int32 and Double providing? Passing those to Number() seems to get the same result. Long will also return the same result but via a different method, but also no less incorrect. I'm looking for a unit test that would fail if one were to change this to `Number(val)` to understand why it is important to fetch the value in this way.

Reply:
Good catch, I didn't realize the checks here aren't necessary since the `valueOf` functions for `BSON.Double` and `BSON.Int32` are just returning `this.value`.

Reply:
I've added in a unit test to flag the Long optimization!